dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/actionlet/SaveContentActionletTest.java

```diff
@@ -83,48 +205,77 @@ private static ContentType createTestType(final ContentTypeAPI contentTypeAPI)
-        contentlet.setProperty("tag", "test");
+        final Contentlet contentlet = new ContentletDataGen(testCase.contentType.id())
+            .setProperty("title", "Test")
+            .setProperty("txt", "Test")
+            .setProperty("tag", "test")
+            .next();
 
-        final Contentlet contentletSaved =
-                workflowAPI.fireContentWorkflow(contentlet,
-                    new ContentletDependencies.Builder()
-                        .modUser(APILocator.systemUser())
-                        .workflowActionId(SystemWorkflowConstants.WORKFLOW_SAVE_ACTION_ID)
-                        .build());
+        Contentlet contentletSaved = null;
 
-        Assert.assertNotNull(contentletSaved);
-        Assert.assertEquals("Test", contentletSaved.getStringProperty("title"));
-        Assert.assertEquals("Test", contentletSaved.getStringProperty("txt"));
+        try {
+            contentletSaved =
+                    workflowAPI.fireContentWorkflow(contentlet,
+                            new ContentletDependencies.Builder()
+                                    .modUser(testCase.user)
+                                    .respectAnonymousPermissions(testCase.respectFrontendRoles)
+                                    .workflowActionId(SystemWorkflowConstants.WORKFLOW_SAVE_ACTION_ID)
+                                    .build());
 
-        final List<TagInode> tagInodes = APILocator.getTagAPI().getTagInodesByInode(contentletSaved.getInode());
-        Assert.assertNotNull(tagInodes);
-        Assert.assertFalse(tagInodes.isEmpty());
+            checkContentSaved(contentletSaved);
+
+            Assert.assertTrue(testCase.hasContentTypeAddChildrenPermission && testCase.hasSaveActionPermission);
+        } catch(Exception e) {
+            if (ExceptionUtil.causedBy(e, DotSecurityException.class)) {
+                Assert.assertTrue(
+                        !testCase.hasContentTypeAddChildrenPermission ||
+                                !testCase.hasSaveActionPermission ||
+                                (systemUser.isFrontendUser() && !testCase.respectFrontendRoles)
+                );
+                return;
+            } else {
+                throw e;
+            }
+        }
 
-        contentletSaved.setTags();
         final Contentlet contentletPublished =
                 workflowAPI.fireContentWorkflow(contentletSaved,
                         new ContentletDependencies.Builder()
                                 .modUser(APILocator.systemUser())
                                 .workflowActionId(SystemWorkflowConstants.WORKFLOW_PUBLISH_ACTION_ID)
+                                .respectAnonymousPermissions(testCase.respectFrontendRoles)
                                 .build());
 
+        checkPublishContent(contentletPublished);
+    }
+
+    private void checkPublishContent(Contentlet contentletPublished) throws DotDataException, DotSecurityException {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'contentletPublished' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5337055)